sql/moz-fx-data-shared-prod/subscription_platform/apple_subscriptions/view.sql

```diff
@@ -0,0 +1,246 @@
+      -- only use most recent expires_date when not limited to status 2
+      ARRAY_AGG(expires_date ORDER BY verified_at DESC LIMIT 1)
+    ) AS end_times,
+  FROM
+    apple_iap_events
+  GROUP BY
+    original_transaction_id
+),
+apple_iap_periods AS (
+  SELECT
+    original_transaction_id,
+    start_time,
+    end_time,
+    period_offset,
+  FROM
+    apple_iap_times
+  LEFT JOIN
+    UNNEST(
+      ARRAY(
+        SELECT AS STRUCT
+          (
+            SELECT
+              MIN(end_time)
+            FROM
+              UNNEST(end_times) AS end_time
+            WHERE
+              end_time > start_time
+          ) AS end_time,
+          MIN(start_time) AS start_time,
+        FROM
+          UNNEST(start_times) AS start_time
+        GROUP BY
+          end_time
+        ORDER BY
+          start_time
+      )
+    )
+    WITH OFFSET AS period_offset
+),
+apple_iap_period_aggregates AS (
+  SELECT
+    periods.original_transaction_id,
+    periods.start_time,
+    periods.end_time,
+    periods.period_offset,
+    COALESCE(
+      MAX(events.legacy_subscription_id),
+      periods.original_transaction_id
+    ) AS original_subscription_id,
+    MIN(events.original_purchase_date) AS original_purchase_date,
+    MAX(events.user_id) AS fxa_uid,
+    ARRAY_AGG(DISTINCT events.offer_identifier IGNORE NULLS) AS promotion_codes,
+    ARRAY_AGG(
+      STRUCT(
+        events.auto_renew_status,
+        events.bundle_id,
+        events.expiration_intent,
+        events.is_in_billing_retry,
+        events.grace_period_expires_date,
+        events.product_id,
+        events.revocation_reason,
+        events.status,
+        events.verified_at
+      )
+      ORDER BY
+        events.verified_at DESC
+      LIMIT
+        1
+    )[OFFSET(0)].*,
+  FROM
+    apple_iap_periods AS periods
+  JOIN
+    apple_iap_events AS events
+  ON
+    periods.original_transaction_id = events.original_transaction_id
+    AND periods.start_time <= events.purchase_date
+    AND periods.end_time > events.purchase_date
+  GROUP BY
+    periods.original_transaction_id,
+    periods.start_time,
+    periods.end_time,
+    periods.period_offset
+),
+apple_iap_enhanced_period_aggregates AS (
+  SELECT
+    *,
+    IF(
+      period_offset > 0,
+      original_subscription_id || "-" || period_offset,
+      original_subscription_id
+    ) AS subscription_id,
+    (
+      (is_in_billing_retry OR grace_period_expires_date > end_time)
+      -- only the last set of active dates can be in the billing grace period
+      AND 1 = ROW_NUMBER() OVER (PARTITION BY original_transaction_id ORDER BY period_offset DESC)
+    ) AS in_billing_grace_period,
+  FROM
+    apple_iap_period_aggregates
+),
+apple_iap_trial_periods AS (
+  SELECT
+    original_transaction_id,
+    purchase_date AS start_time,
+    MAX(expires_date) AS end_time,
+  FROM
+    apple_iap_events
+  WHERE
+    offer_type = 1
+  GROUP BY
+    original_transaction_id,
+    purchase_date
+)
```

Comment:
Because we can't join on `start_time` we need to make sure we only get one trial per `original_transaction_id`.
```suggestion
apple_iap_trial_periods AS (
  SELECT
    original_transaction_id,
    purchase_date AS start_time,
    expires_date AS end_time,
  FROM
    apple_iap_events
  WHERE
    offer_type = 1
  QUALIFY
    1 = ROW_NUMBER() OVER (PARTITION BY original_transaction_id ORDER BY expires_date DESC)
)
```